File: playlist.js
const express = require("express");
const cron = require("node-cron");
const bodyParser = require("body-parser");


const app = express();
app.use(bodyParser.json());

let playlist = [];

// schedule song playback
function ScheduleSong(song){
    const{name, time} = song;
    const task =cron.schedule( time, ()=> {
        console.log(`now Playing: ${name}`);
    });

    return task;
}

// Route to add a song to the playlist
app.post("/add", (req, res) => {
    const {name, time} = req.body;
    if (!name || ! time){
        return res.status(400).json({error: "Song name and time required"});

    }

    const task = scheduleSong({name, time});
    playlist.push({name, time, task});

    res.json({message: "Song Scheduled", name, time});
});


// Route to list all scheduled songs
app.get("/playlist", (req, res) => {
    res.json(playlist.map(song => ({ name: song.name, time: song.time })));
}); 


// Route to remove a song
app.delete("/remove/:name", (req, res) => {
    const name = req.params.name;
    const index = playlist.findIndex(song => song.name === name);

    if (index === -1) {
        return res.status(404).json({ error: "Song not found!" });
    }

    playlist[index].task.stop();
    playlist.splice(index, 1);
    res.json({ message: `Song ${name} removed.` });
});

const PORT = 3000;
app.listen(PORT, () => console.log(`Playlist Scheduler running on port ${PORT}`));

File: Python/playlist.py
from flask import Flask, request, jsonify
from apscheduler.schedulers.background import BackgroundScheduler
import time

app = Flask(__name__)
scheduler = BackgroundScheduler()
scheduler.start()

playlist = {}

# Function to play a song
def play_song(name):
    print(f" Now playing:{name}")


# Route to add a song to the playlist
@app.route("/add", methods = ["POST"])
def add_song():
    data = request.get_json()
    name = data.get("name")
    time_delay = data.get("time") # Delay in seconds

    if not name in time_delay:
        return jsonify({"error": "Song name and delay time required"}), 400
    job = scheduler.add_job(play_song, 'interval', seconds =int(time_delay), id=name, args=[name])
    playlist[name] = job
    return jsonify({"message": "Song added to playlist!", "name": name, "interval": time_delay })

# Route to list all the scheduled songs
@ app.route("/playlist", methods = ["GET"])
def list_song():
    return jsonify(list(playlist.keys()))

# Route to remove a song from the playlist
@app.route("/remove/<name>", methods = ["DELETE"])

def remove_song(name):
    if name in playlist:
        playlist[name].remove()
        del playlist[name]
        return jsonify({"message": f"Song{name} removed from playlist."})
    return jsonify({"error": f"Song {name} not found!"}), 404

if __name__ == "__main__":
    app.run(debug = True, port=5000)
